Verify transfer spl-token on solana, supporting multi chain solana

mohammadranjbarz opened the pull request:
related to #1175

@CarlosQ96  @aminlatifi  I changed something about networks and chains to support dev, test and mainnet in Solana

mohammadranjbarz committed:
Verify transfer spl-token on solana, supporting multi chain solana

related to #1175

CarlosQ96 committed:
fix solana tests on devnet

CarlosQ96 commented:
Pushed graphql validators fix and network error. 
Will Push migration for other tokens

CarlosQ96 committed:
add solana additional tokens

CarlosQ96 committed:
fix solana token transfer test

CarlosQ96 commented:
I checked testnet and devnet but I couldn't find some tokens. 

Some tokens have multiple different test ones, like USDc dev and USDC coin dev, usdc saber testnet.

Should we keep the mainnet ones only?

CarlosQ96 reviewed (approved):
Thanks Mohammad, your code look good to me. Can review mine too now jajaja

mohammadranjbarz commented on src/utils/validators/graphqlQueryValidators.ts:
```diff
@@ -99,9 +99,17 @@ export const createDonationQueryValidator = Joi.object({
     .valid(...Object.values(NETWORK_IDS)),
   tokenAddress: Joi.when('transactionNetworkId', {
-    is: 0, // if its solana network
+    is: NETWORK_IDS.SOLANA_MAINNET,
```
@CarlosQ96  IT doesnt need to add this exception for SOLANA DEVNET and TESTNET as well?

CarlosQ96 replied:
Its in the otherwise clauses.

mohammadranjbarz replied:
Oh you are right

aminlatifi replied:
First, I think there would be a better solution than nesting, like 
```suggestion
    is: Joi.valid(NETWORK_IDS.SOLANA_MAINNET, NETWORK_IDS.SOLANA_DEVNET, NETWORK_IDS.SOLANA_TESTNET),
```

Beyond that, I guess it's better to decide the valid regext to match based on chainType rather than networkId

CarlosQ96 replied:
Done

aminlatifi commented on src/provider.ts:
```diff
@@ -18,7 +18,11 @@ export const NETWORK_IDS = {
   CELO_ALFAJORES: 44787,
   ETC: 61,
   MORDOR_ETC_TESTNET: 63,
-  SOLANA: 0,
```
The SOL token was by migration `migration/1703044586989-addSolanaToken.ts` before this change, will have networkId=0. We must address that

CarlosQ96 replied:
I think we can address it manually, the migration uses the seedtoken file which was updated, so deploying to develop or production will have the correct values.

Ill address it manually in staging.

mohammadranjbarz commented:
Thanks @CarlosQ96  
Your codes look good to me.

CarlosQ96 committed:
improve joi schema chaintype validator

aminlatifi commented on src/utils/validators/graphqlQueryValidators.ts:
```diff
@@ -98,18 +99,10 @@ export const createDonationQueryValidator = Joi.object({
-  tokenAddress: Joi.when('transactionNetworkId', {
-    is: NETWORK_IDS.SOLANA_MAINNET,
+  tokenAddress: Joi.when('chainType', {
+    is: Joi.string().pattern(chainTypeRegex),
```
It must match only with solana

CarlosQ96 replied:
Whoops I pushed wrong regex during testing. Fixed.

aminlatifi replied:
So, why regex? :)
It only would match with "SOLANA" value which can be simple equal condition (then remove `chainTypeRegex` variable). Also, it's better to use ChainType.SOLANA instead of repeating "SOLANA" here.

mohammadranjbarz replied:
I changed it, done.

aminlatifi commented:
@mohammadranjbarz does this resolve https://github.com/Giveth/impact-graph/issues/1172 as well?

mohammadranjbarz commented:
> @mohammadranjbarz does this resolve #1172 as well?

Yeah we check the token account for each SOLANA wallet, so yes I think it's handled

aminlatifi commented on src/services/chains/solana/transactionService.ts:
```diff
@@ -10,16 +10,33 @@ import {
+    solanaProviders[chainId] = new SolanaWeb3.Connection(
+      process.env.SOLANA_DEVNET_NODE_RPC_URL as string,
+    );
+  }
```
let's benefit from switch-case for more readable code.

mohammadranjbarz replied:
@aminlatifi  Yes you are right, Done

aminlatifi commented on src/services/chains/solana/transactionService.ts:
```diff
@@ -62,10 +79,55 @@ export const getSolanaTransactionDetailForSolanaTransfer = async (
+
+      // we already check toAddressBalance.owner === params.toAddress
+      to: toAddressBalance?.owner,
+      amount: txInfo.tokenAmount?.uiAmount,
```
The example I have doesn't hold any value in parsed.info.tokenAmount.uiAmount!
The safer approach would be get receivers balance in preTokenBalances and postTokenBalances, then use their diff here.

aminlatifi replied:
I got my example by running this
```ts
 const connection = new Connection(clusterApiUrl("devnet"), "confirmed");
  const txInfo = await connection.getParsedTransaction(
    "3m6f1g2YK6jtbfVfuYsfDbhVzNAqozF8JJyjp1VuFDduecojqeCVK4htKnLTSk3qBwSqYUvgLpBTVpeLJRvNmeTg"
  );
  console.log(JSON.stringify(txInfo, null, 2));
  ```

mohammadranjbarz replied:
let me write a test case for your example, to see what happens

aminlatifi replied:
Some info about the transfer
From wallet: 26Aks2rN6mfqxdYRXKZbn8CS4GBv6fCMGFYfGWvfFfcx
To wallet: 7TJgw4hDHh5wdKep3EsBkGMSvtf9LsxdXf89LA48uHoq
Mint address: "8LDBhHJB7oMAjkJaetXa4njjetUVWDRTqvzkmhFQjgeK"
From Token Account: 9VsnTMDpnSoMhYx2FQpXAs5f4a1UNh9WxUBLL6LpiYZo
To Token Account: HhYuUyc9T7V758kfR3f8PG8pKz38MsDvhkQtEEtbVBrV
Mint tx: 5NDEXP9F8UncZqZsMcssSRjX4YMHEz2iZyDgfCcPdSjzcAP1u8Kg2Js7BCKz4taiLSsK88DqdiFnCEZZLfyo1fpg
Transfer tx: 3m6f1g2YK6jtbfVfuYsfDbhVzNAqozF8JJyjp1VuFDduecojqeCVK4htKnLTSk3qBwSqYUvgLpBTVpeLJRvNmeTg

mohammadranjbarz replied:
Thanks I'm working on it

mohammadranjbarz committed:
Add another spl-token transfer test case, change validator of createDonation webservice

related to https://github.com/Giveth/impact-graph/pull/1232#discussion_r1444472435

mohammadranjbarz committed:
Refactor and change a nested if-else with witch case

aminlatifi committed:
Removed only tags from tests

aminlatifi commented on src/utils/locales/en.json:
```diff
@@ -1,105 +1,107 @@
+	"REGISTERED_NON_PROFITS_CATEGORY_DOESNT_EXIST": "There is not any category with name registered-non-profits, probably you forgot to run migrations",
+	"PROJECT_UPDATE_CONTENT_LENGTH_SIZE_EXCEEDED": "Content length exceeded",
+	"INVALID_TOKEN_ADDRESS": "INVALID_TOKEN_ADDRESS",
+	"Invalid tokenAddress": "Invalid tokenAddress"
```
```suggestion
	"INVALID_TOKEN_ADDRESS": "Invalid tokenAddress"
```

aminlatifi commented on src/utils/errorMessages.ts:
```diff
@@ -231,6 +232,7 @@ export const translationErrorMessagesKeys = {
     'CATEGORIES_MUST_BE_FROM_THE_FRONTEND_SUBSELECTION',
   INVALID_TX_HASH: 'INVALID_TX_HASH',
   INVALID_TRANSACTION_ID: 'INVALID_TRANSACTION_ID',
+  INVALID_TOKEN_ADDRESS: 'Invalid tokenAddress',
```
```suggestion
  INVALID_TOKEN_ADDRESS: 'INVALID_TOKEN_ADDRESS',
```

CarlosQ96 committed:
Update src/utils/errorMessages.ts

Co-authored-by: Amin Latifi <a.latifi.al@gmail.com>

CarlosQ96 committed:
Update src/utils/locales/en.json

Co-authored-by: Amin Latifi <a.latifi.al@gmail.com>
